presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java

```diff
@@ -3125,6 +3185,96 @@ protected void doTestTransactionDeleteInsert(HiveStorageFormat storageFormat, Sc
+
+        doCreateEmptyTable(schemaTableName, storageFormat, MISMATCH_SCHEMA_TABLE_BEFORE);
+
+        // insert the data
+        try (Transaction transaction = newTransaction()) {
+            ConnectorSession session = newSession();
+            ConnectorMetadata metadata = transaction.getMetadata();
+            ConnectorTableHandle tableHandle = getTableHandle(metadata, schemaTableName);
+
+            ConnectorInsertTableHandle insertTableHandle = metadata.beginInsert(session, tableHandle);
+            ConnectorPageSink sink = pageSinkProvider.createPageSink(transaction.getTransactionHandle(), session, insertTableHandle);
+            sink.appendPage(MISMATCH_SCHEMA_TABLE_DATA_BEFORE.toPage(), null);
+            Collection<Slice> fragments = sink.finish();
+
+            metadata.finishInsert(session, insertTableHandle, fragments);
+
+            transaction.commit();
+        }
+
+        // load the table and verify the data
+        try (Transaction transaction = newTransaction()) {
+            ConnectorSession session = newSession();
+            ConnectorMetadata metadata = transaction.getMetadata();
+            ConnectorTableHandle tableHandle = getTableHandle(metadata, schemaTableName);
+
+            List<ColumnHandle> columnHandles = metadata.getColumnHandles(session, tableHandle).values().stream()
+                    .filter(columnHandle -> !((HiveColumnHandle) columnHandle).isHidden())
+                    .collect(toList());
+
+            MaterializedResult result = readTable(transaction, tableHandle, columnHandles, session, TupleDomain.all(), OptionalInt.empty(), Optional.empty());
+            assertEqualsIgnoreOrder(result.getMaterializedRows(), MISMATCH_SCHEMA_TABLE_DATA_BEFORE.getMaterializedRows());
+            transaction.commit();
+        }
+
+        // alter the table schema
+        try (Transaction transaction = newTransaction()) {
+            ConnectorSession session = newSession();
+            PrivilegeGrantInfo allPrivileges = new PrivilegeGrantInfo("all", 0, session.getUser(), PrincipalType.USER, true);
+            PrincipalPrivilegeSet principalPrivilegeSet = new PrincipalPrivilegeSet(
+                    ImmutableMap.of(session.getUser(), ImmutableList.of(allPrivileges)),
+                    ImmutableMap.of(),
+                    ImmutableMap.of());
+            Table oldTable = transaction.getMetastore(schemaName).getTable(schemaName, tableName).get();
+            HiveTypeTranslator hiveTypeTranslator = new HiveTypeTranslator();
+            List<Column> dataColumns = MISMATCH_SCHEMA_TABLE_AFTER.stream()
+                    .filter(columnMetadata -> !columnMetadata.getName().equals("ds"))
+                    .map(columnMetadata -> new Column(columnMetadata.getName(), toHiveType(hiveTypeTranslator, columnMetadata.getType()), Optional.empty()))
+                    .collect(toList());
+            Table.Builder newTable = Table.builder()
```

Comment:
I believe you can drop most of the following if you initialize the builder with the table.  Like this: `Table.Builder newTable = Table.builder(oldTable)`